### PR TITLE
Add Jaeger support

### PR DESCRIPTION
**Pull request by andriylesch**

### CODE REVIEW

File: examples/basic/main.go [hunk 1]
```diff
@@ -13,7 +13,11 @@ const (
 )
 
 func main() {
-	tracing.SetGlobalTracer(appName, "http://localhost:9411")
+	err := tracing.SetGlobalTracer(appName)
+	if err != nil {
+		panic(err)
```
Comment (FrancoisPoinsot) [inline examples/basic/main.go]: don't panic!

**Comment by cmeury** [comment]
Pushed docker-compose configuration, would be great if you could have a look.

File: Makefile [hunk 2]
```diff
@@ -1,7 +1,15 @@
...
 
 .PHONY: test
 test:
-	go test -v `go list ./... | grep -v /examples/`
\ No newline at end of file
+	go test -v `go list ./... | grep -v /examples/`
+
+
+docker-compose:
+	@CGO_ENABLED=0 GOOS=linux go build -o ./app -a -ldflags '-s' -installsuffix cgo examples/$(EXAMPLE_APP)/main.go
```
Comment (FrancoisPoinsot) [inline Makefile]: why having a make step that only run one example?

File: README.md [hunk 3]
```diff
@@ -40,6 +40,15 @@ To watch traces you just have to hit http://localhost:9411/
  - Extract/Inject span from/to a map[string]string (textMapCarrier)
  - Declare an error span
 
+## Local Testing
+
+To locally test the tracer for development, use the provided `docker-compose` configuration to spin up jaeger and
```
Comment (FrancoisPoinsot) [inline README.md]: all examples were already describe in `#example` section

File: docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      - "6831:6831"   # UDP - accept jaeger.thrift over compact thrift protocol
+      - "6832:6832"   # UDP - accept jaeger.thrift over binary thrift protocol
+      - "5778:5778"   # HTTP - serve configs
+      - "16686:16686" # HTTP - serve frontend
+      - "14268:14268" # HTTP - accept jaeger.thrift directly from clients
+      - "9411:9411"   # HTTP - Zipkin compatible endpoint
+    environment:
+      - COLLECTOR_ZIPKIN_HTTP_PORT=9411
+
+#  prometheus:
```
Comment (FrancoisPoinsot) [inline docker-compose.yml]: dead config?

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,57 @@
+version: '3'
+
+services:
+  api:
+    build: .
```
Comment (FrancoisPoinsot) [inline docker-compose.yml]: let's not add the app itself to the docker. Not like there is any single app, this is a library
pro:
- you can debug your application
- you can start any example you want

cons:
- you need 2 clicks now

File: wait-for-it.sh [hunk 6]
```diff
@@ -0,0 +1,79 @@
+#!/bin/sh
```
Comment (FrancoisPoinsot) [inline wait-for-it.sh]: i am all for open sourcing and all but here is the same on 4 lines:

```sh
@until $(curl --output /dev/null --silent --head --fail http://localhost:5775); do \
		printf '.'; \
		sleep 2; \
	done
```

maybe it is less generic but i would rather use that

Comment (FrancoisPoinsot) [reply wait-for-it.sh]: or download the original as a dependency

File: tracer.go [hunk 7]
```diff
@@ -1,73 +1,37 @@
...
-	collector, err := zipkin.NewHTTPCollector(absoluteEndPointURL.String())
-
+func SetGlobalTracer() error {
+	cfg, err := jaegercfg.FromEnv()
 	if err != nil {
-		return nil, errors.Wrap(err, "createCollector")
+		return err
+	} else {
+		var tracer opentracing.Tracer
+		tracer, globalCloser, err = cfg.New(
```
Comment (FrancoisPoinsot) [inline tracer.go]: deprecated

**Comment by FrancoisPoinsot** [comment]
This repo is public
We already received some PR from external guys using the zipkin implementation.
It didn't take to much not to drop it completly, but default to jaeger.

This remain a breaking change for anyone who wish to use zipkin: they need to pass option `UsingZipkin` .

With this middle ground I hope we continue to receive PR from random guy on the internet. :)

File: tracer.go [hunk 8]
```diff
@@ -1,73 +1,77 @@
...
-	)
+func SetGlobalTracer(options ...TracerOption) error {
+	config := &TracerConfig{}
+	for _, option := range options {
+		option(config)
+	}
 
-	if err != nil {
-		return nil, errors.Wrap(err, "createTracer")
+	switch config.TracingService {
```
Comment (kazi-ricardo) [inline tracer.go:45]: why not simply `if / else`?

File: tracer.go [hunk 9]
```diff
@@ -1,73 +1,77 @@
...
-		return errors.Wrap(err, "SetGlobalTracer")
+type TracerConfig struct {
+	TracingService string
+	AppName        string
+	TracingHost    string
+}
+
+type TracerOption func(config *TracerConfig)
+
+func UsingZipkin(appName string, host string) TracerOption {
```
Comment (kazi-ricardo) [inline tracer.go]: as this is a shared library some docs/comments would be appreciated

File: README.md [hunk 10]
```diff
@@ -23,14 +23,14 @@ defer span.Finish()
...
 go run examples/basic/main.go
 go run examples/httpServer/main.go
 go run examples/httpServer-middleware/main.go
 go run examples/httpGoKit-middleware/main.go
 ```
 
-To watch traces you just have to hit http://localhost:9411/
+To watch traces you just have to hit http://localhost:16686/search
 
 ## Features
```
Comment (kazi-ricardo) [inline README.md:38]: Info about required ENVs for `jaeger` would be useful

File: tracer.go (hunk 8 above)
Comment (FrancoisPoinsot) [reply tracer.go:45]: just felt like it

File: README.md (hunk 10 above)
Comment (FrancoisPoinsot) [reply README.md:38]: they are part of the examples

Comment (FrancoisPoinsot) [reply README.md:38]: will add more description

**Review by cmeury — APPROVED** [review]

**Pull request merged**